docs/PSEUDOCODE.md

```diff
@@ -0,0 +1,191 @@
+that have already been processed.
+
+### Log Trigger Postprocessing
+Log trigger postprocessing is a combination of putting eligible results into
+OCR staging and sending retryable results to the retry ticker. The last part is
+unique to log triggers and is not applicable to conditionals.
+```
+for each check pipeline result
+    if check pipeline result is eligible
+        add result to ocr staging
+    else if check pipeline result is retryable
+        add check result to retry ticker
+```
+
+### Conditional Preprocessing
+Conditional upkeeps need to be continuously processed at progressing blocks. A
+single upkeep can either be in-flight or transmitted as a stale report. OCR
+reporting blocks upkeeps and chain events unblock them each ensuring that the
+next perform will happen at on a block in the future.
+```
+for each upkeep payload
+    if upkeep is pending or stale
+        remove upkeep payload from provided list
+
+return upkeep payload list
+```
+
+### Conditional Postprocessing
+Contitional upkeeps require previous results that have not been processed to be
+replaced by newly created results to ensure block numbers and results are kept
+fresh.
+```
+remove results from previous tick from ocr staging
+
+for each check pipeline result
+    if check pipeline result is eligible
+        add result to ocr staging
+```
+
+### Retry Postprocessing
+Retries happen when errors are surfaced through the check pipeline. An upkeep
+needs to be sent back through the check pipeline at a future time based on some
+amount of timeout. Consider exponential backoff per upkeep as an example delay
+mechanism.
+
+A retry component can be structured as a combination of a ticker and 
+postprocessor hook. Postprocessing provides data to the ticker and the ticker
+runs an observer that pushes values into OCR staging.
+```
+[postprocessor]
+for each check result
+    if check result is retryable
+        add to retry ticker
+
+[retry ticks]
+for time interval
+    for each retryable upkeep payload
+        if upkeep is retryable now
+            add to next retries
+
+send retries as tick
+```
+
+### Conditional Sampling
+To reduce over all check load on the nodes for conditional upkeeps, each node
+only checks a random subset of upkeeps per tick interval. The tick interval is
+derived by block cadence, but not every chain will have every block checked.
+Fast chains will have multiple blocks skipped while sampling takes place.
+```
+
+```
+
+## OCR3 Report Flow
+OCR is allowed to run on a separate cadence from the eligibility check flow. In
+general, it pulls from some collection of staged values, constructs observations,
+and builds reports. Some components hook into this process to feed progress
+data to the eligibility check flows.
+
+For the case of node restarts, it was proposed that local caches would be
+cleared exposing a possibility that upkeeps could be reported by a set of 
+restarted nodes resulting in successful (bad) transmits.
```

Comment:
This should be fine for now, the contract can protect against duplicated transmits